docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+
+## Goals
+
+This framework design should strive to accomplish the following with validation support.
+
+Allow provider developers access to all current types of provider validation:
+
+- Provider configurations
+- Resource configurations (and configurations during plans)
+- Data Source configurations (and configurations during plans)
+
+Including where possible:
+
+- Single attribute value validation (e.g. string length)
+- Multiple attribute validation (e.g. attributes or attribute values that conflict with each other)
+
+In terms of implementation, the following core concepts:
+
+- Low level primitives (e.g other portions of the framework, external packages, and provider developers can implement higher level functionality)
+- Reusability between single attribute and multiple attribute validation functionality (e.g. attribute value functions)
+- Hooks for documentation (e.g. for future tooling such as provider documentation generators to self-document attributes)
+
+Finally, these other considerations:
+
+- Providing the appropriate amount of contextual information for debugging purposes (e.g. logging)
+- Providing the appropriate amount of contextual information for practitioner facing output (e.g. paths and values involved with validation decisions)
+- Ease of extending validation (e.g. handling type conversion and/or unknown values in the framework)
+- Ease of testing validation (e.g. unit testing)
+- Ease and succinctness of common validation scenarios (e.g. verbosity in provider code)
+- Allowing potential future enhancements of validation behavioral decisions based on configuration (e.g. converting validation errors to warnings or logs)
+
+## Proposals
+
+### Extending Plan Modifications Versus New Handling
+
+#### Extending Plan Modifications
+
+The [Plan Modifications design documentation](./plan-modifications.md) outlines proposals which broadly replace the previous framework's `CustomizeDiff` functionality. See that documentation for considerations and recommendations there. In this proposal for validation, new functions for validation would be provided within that framework, rather than introducing separate handling.
+
+Implementing against that design could prove complex for the framework as they are intended to serve differing purposes. It could also be confusing for provider developers in the same way that `CustomizeDiff` was confusing where differing logical rules applied to differing attribute value and operation scenarios. Another wrinkle is that plan modifications are only intended to run during `terraform plan` (`PlanResourceChanges` RPC) and `terraform apply` (`ApplyResourceChanges` RPC), so the framework would be introducing its own additional logic to extract and perform any validation functions during the `terraform validate` (`ValidateDataSourceConfig`/`ValidateProviderConfig`/`ValidateResourceConfig` RPCs).
+
+#### New Handling
```

Comment:
"New Handling" confused me for a minute. Is something like "New Abstraction" clearer? 🤔

Reply:
Love it, will be included in next commit.